### ISSUE TITLE
Do you have any thoughts about reforming YOLOV5?

### ISSUE THREAD

**Issue by jamjamjon** [ultralytics/yolov5#5495]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

Do you have an idea to make a big modify with yolov5?  Like changing backbone,  detect head, even removing the anchors like YOLOX. YOLOV5 is powerful right now, But I just want to see some big change in a short future.  If you have any thoughts about it, please let me know. I really like your work in YOLOV5! Have a good day!

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by zhiqwang** [comment]
IMO, The paradigm currently adopted by YOLOv5 is very refined, and this approach is also highly scalable and configurable. On the other hand, you may also see some refactored versions of yolov5 in the GitHub community with the `backbone`, `fpn(PAN)` and `head` forms, such as the following 

- https://github.com/zhiqwang/yolov5-rt-stack
- https://github.com/Okery/YOLOv5-PyTorch
- https://github.com/yl305237731/flexible-yolov5

**EDIT:** I think I've misunderstood this ticket, just ignore my comment above :/

**Comment by glenn-jocher** [comment]
@Jamjamjamjon yes on the architecture side there is of course a lot of experimentation we could do, but a lot of this is high risk basic research that may not produce improvements.

BUT, there is a lot of low risk development also we need to do, mostly related to MLops/Devops capabilities that should produce benefits for everyone:
- Single and Multi-GPU CI
- Major code refactor to create 1) multi-backend (i.e. PyTorch, TF, Keras, TFLite, CoreML, torchscript) model loading, 2) multi-backend inference functions. This would allow detect.py and val.py to share model loading and inference functions, and then this would allow val.py to test exported models (TFLite, ONNX, CoreML etc.), which multiple users have been asking for. This would really help users verify that their deployed models are performing well, they could quantify drops in performance from quantization etc. which we don't offer now.
- Better introspection tools in val.py to highlight problems, i.e. edge cases, poorly performing images, poorly performing classes. This would help inform dataset updates to target weak points. For example in COCO some classes perform very poorly, and I'm sure some images in particular perform poorly, we want to be able to display those problem images, which we can't do now since metrics are not computed per image.

EDIT: Another big point is we'd like to expand the tasks we target to classification and segmentation. With classification we've started a branch at https://github.com/ultralytics/yolov5/blob/classifier/classifier.py, with segmentation we still have a lot more ways to go (new segmentation head, loss function, plots/annotation updates).

**Comment by jamjamjon** [comment]
thanks！I will make some change with model architecture and do some experiments. I will let you know once I get something. Thanks again！Have a good day！



发自我的iPhone


------------------ Original ------------------
From: Glenn Jocher ***@***.***&gt;
Date: Thu, Nov 4, 2021 8:05 PM
To: ultralytics/yolov5 ***@***.***&gt;
Cc: Jamjamjamjon ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [ultralytics/yolov5] Do you have any thoughts about reformingYOLOV5? (Issue #5495)





 
@Jamjamjamjon yes on the architecture side there is of course a lot of experimentation we could do, but a lot of this is high risk basic research that may not produce improvements.
 
BUT, there is a lot of low risk development also we need to do, mostly related to MLops/Devops capabilities that should produce benefits for everyone:
  
Single and Multi-GPU CI 

Major code refactor to create 1) multi-backend model loading, 2) multi-backend inference functions. This would allow detect.py and val.py to share model loading and inference functions, and then this would allow val.py to test exported models (TFLite, ONNX, CoreML etc.), which multiple users have been asking for. 

Better introspection tools in val.py to highlight problems, i.e. edge cases, poorly performing images, poorly performing classes. This would help inform dataset updates to target weak points. For example in COCO some classes perform very poorly, and I'm sure some images in particular perform poorly, we want to be able to display those problem images, which we can't do now since metrics are not computed per image. 
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by glenn-jocher** [comment]
@Jamjamjamjon great! Yes please help future uses by submitting PRs for any improvements or bug fixes you come across.
